contracts/orderbook/Logic/OrderBook.sol

```diff
@@ -0,0 +1,413 @@
+        address start,
+        address end,
+        uint256 amount
+    ) public view returns (uint256) {
+        (uint256 executionPrice, uint256 precision) = IPriceFeeds(getFeed())
+            .queryRate(start, end);
+        return (executionPrice * amount) / precision;
+    }
+
+    function _prepDexAndPayload(bytes memory input)
+        internal
+        pure
+        returns (uint256 dexID, bytes memory payload)
+    {
+        if (input.length != 0) {
+            (uint128 flag, bytes[] memory payloads) = abi.decode(
+                input,
+                (uint128, bytes[])
+            );
+            if(flag & DEX_SELECTOR_FLAG != 0){
+                (dexID, payload) = abi.decode(payloads[0], (uint256, bytes));
+            }
+        }
+    }
+
+    function prelimCheck(bytes32 orderID) external returns (bool) {
+        IOrderBook.Order memory order = _allOrders[orderID];
+        uint256 amountUsed;
+        address srcToken;
+        (uint256 dexID, bytes memory payload) = _prepDexAndPayload(order.loanDataBytes);
+        if (payload.length == 0) {
+            if (order.orderType == IOrderBook.OrderType.LIMIT_OPEN) {
+                payload = abi.encode(order.loanTokenAddress, order.base);
+            } else {
+                payload = abi.encode(order.base, order.loanTokenAddress);
+            }
+            dexID = 1;
+        }
+        ISwapsImpl swapImpl = ISwapsImpl(
+            IDexRecords(getSwapAddress()).retrieveDexAddress(dexID)
+        );
+        srcToken = order.collateralTokenAmount > order.loanTokenAmount ? order.base : order.loanTokenAddress;
+        if (order.timeTillExpiration < block.timestamp) {
+            return false;
+        }
+        if (order.orderType == IOrderBook.OrderType.LIMIT_OPEN) {
+            if (order.loanID != 0 && !_isActiveLoan(order.loanID)) {
+                return false;
+            }
+            uint256 dSwapValue;
+            if (srcToken == order.loanTokenAddress) {
+                amountUsed = order.loanTokenAmount + (order.loanTokenAmount * order.leverage) / 10**18; //adjusts leverage
+                (dSwapValue, ) = swapImpl.dexAmountOutFormatted(
+                    payload,
+                    amountUsed
+                );
+            } else {
+                amountUsed = queryRateReturn(
+                    order.base,
+                    order.loanTokenAddress,
+                    order.collateralTokenAmount
+                );
+                amountUsed = (amountUsed * order.leverage) / 10**18;
+                (dSwapValue, ) = swapImpl.dexAmountOutFormatted(
+                    payload,
+                    amountUsed
+                );
+                dSwapValue += order.collateralTokenAmount;
+            }
+
+
+            if (order.amountReceived <= dSwapValue) {
+                return true;
+            }
+        } else if (order.orderType == IOrderBook.OrderType.LIMIT_CLOSE) {
+            if (!_isActiveLoan(order.loanID)) {
```

Comment:
you can incorporate this IF in the IF above

Reply:
true. I need to re-work the else statement though but it should be more efficient.